### PR TITLE
dev-cmd/bottle: use system GNU tar if possible

### PR DESCRIPTION
**Pull request by cho-m**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

I switched to non-reproducible args for building `gnu-tar` bottle in #18138 due to problems bottling `gnu-tar` with `gnu-tar`.

Looking into this again to restore reproducibility. Incomplete PR as back-to-back bottles are still producing a diff on Linux.

---

The main issue for `gnu-tar` is running with placeholders, particularly for the interpreter.

On Linux, there are 2 options:
1. Use system GNU tar (changes in this PR). Potential limitations are if someone on old Linux tries bottling `gnu-tar` (unlikely) and maybe the version number difference.
2. Directly run interpreter, e.g.
    ```
    #{HOMEBREW_PREFIX}/lib/ld.so \
      --library-path #{HOMEBREW_PREFIX}/lib:... \
      /path/to/tar ...
    ```

On macOS, `gnu-tar` has no dependencies and doesn't have interpreter handling like Linux so can just bottle `gnu-tar` with `gnu-tar`:
```console
❯ brew bottle --json --only-json-tab gnu-tar -v
/opt/homebrew/Library/Homebrew/vendor/bundle/ruby/3.3.0/bin/bundle clean
==> Determining gnu-tar bottle rebuild...
==> Bottling gnu-tar--1.35.arm64_sequoia.bottle.2.tar.gz...
==> Installing `gnu-tar` for bottling...
/opt/homebrew/bin/brew install --formula gnu-tar
Warning: gnu-tar 1.35 is already installed and up-to-date.
To reinstall 1.35, run:
  brew reinstall gnu-tar
/opt/homebrew/opt/gnu-tar/bin/gtar --create --numeric-owner --mtime=2023-07-18 02:55:23 --sort=name --owner=0 --group=0 --numeric-owner --format=pax --pax-option=globexthdr.name=/GlobalHead.%n,exthdr.name=%d/PaxHeaders/%f,delete=atime,delete=ctime --file /Users/cho-m/Desktop/gnu-tar--1.35.arm64_sequoia.bottle.2.tar gnu-tar/1.35
./gnu-tar--1.35.arm64_sequoia.bottle.2.tar.gz
  bottle do
    rebuild 2
    sha256 cellar: :any_skip_relocation, arm64_sequoia: "74c997076b93549ba3b8a5bd59d768a853b929d33317cc59ba7ec448824eff6f"
  end
Writing gnu-tar--1.35.arm64_sequoia.bottle.json
```

### CODE REVIEW

**Comment by Bo98** [comment]
I believe the reason for using Homebrew `gnu-tar` was because there was output differences in a recent version and the system tar being older caused issues. https://github.com/Homebrew/brew/commit/71888db8ba1398287a867ea068f97fd8fb6380ec

**Comment by MikeMcQuaid** [comment]
> Makes sense to use system tar when bottling gnu-tar itself though.

Agreed 👍🏻

**Comment by Bo98** [comment]
The compression algorithm is really a completely different thing. `tar` handles it for ease of use but under the hood it doesn't actually implement any compression code and simply defers to another command so I wouldn't really conflate our choice of `tar` for future compression support. `tar` does save having to remember what those external commands are but you still need to handle making sure they're installed etc anyway and handle the different media types in the OCI manifest.

Using Ruby Gzip has also been a good improvement as we don't need to worry about the system so that does set a fairly high quality bar.

**Comment by cho-m** [comment]
> Would only want to do this if we can do so for all supported macOS versions (and typical Ubuntu runners) with no extra Homebrew or system dependencies.

Probably won't be possible for all formulae unless Apple decides to ship a copy.

If only handling a subset of formulae, then possible if `zstd` is in dependency tree already, e.g. anything that depends on `libtiff`.

For performance, we may need to look into relocation handling first.

---

> Using Ruby Gzip has also been a good improvement as we don't need to worry about the system anymore (particularly GNU Gzip's different implementation) so that does set a fairly high quality bar, whereas meanwhile we've had to adjust this `tar` code repeatedly.

Letting `tar` handle both phases does allow some small optimizations for intermediary files. But avoiding implementation details (particularly dealing with oldest macOS/Linux support) does help.

On side note, using equivalent of `gzip --no-name` would probably have been ideal for reproducible builds but probably too late for that since our existing bottles have name/timestamp.

**Comment by MikeMcQuaid** [comment]
> Probably won't be possible for all formulae unless Apple decides to ship a copy.

Yes, I consider this the requirement here.

**Comment by Bo98** [comment]
One problem with `zstd` is it is not very reproducible. It changes between versions and, in certain cases, [between CPUs](https://redirect.github.com/facebook/zstd/issues/4099).

I know we've broken reproducibility a lot in the last year, though I think it's still something we aim to keep fairly stable.

On macOS, we do have options like LZFSE but I suspect it's not too big of an impact.

**Comment by cho-m** [comment]
> One problem with `zstd` is it is not very reproducible. It changes between versions and, in certain cases, [between CPUs](https://redirect.github.com/facebook/zstd/issues/4099).

Curious what Ubuntu, Fedora and Arch Linux do here given they use `zstd`, though haven't checked their stance on reproducibility. The ones focused on reproducibility like Debian and NixOS currently use `xz`.

Though either way not possible right now in Homebrew/core. macOS tar can only decompress `.gz`, `.bz2`, or `.xz` without extra dependencies and the others are all worse at compression/decompression speed.

---

> On macOS, we do have options like LZFSE but I suspect it's not too big of an impact.

Seems like it would require more effort/code to use. And attempting `all` bottles would require more work. I guess it may have better decompression speed though doesn't seem worth it.

---

Anyway, will probably get back to PR soon.

Still looking at best place to handle logic, i.e. whether worth moving to Linux-specific code or keeping the GNU tar check and just adding a `formula.name == "gnu-tar"`.

**Comment by Bo98** [comment]
> Curious what Ubuntu, Fedora and Arch Linux do here given they use `zstd`, though haven't checked their stance on reproducibility. The ones focused on reproducibility like Debian and NixOS currently use `xz`.

Given the branched approach to many distros, they might not care too much about reproducibility between branches but do within a branch. I don't think zstd patch releases ever get backported - only security update commits.

For Homebrew, we're rolling release so we don't really have that distinction and break as an when it makes sense to but ultimately try to have some reproducibility.

**Comment by carlocab** [comment]
> I do agree zstd is ideal, though any global dependency on a formula always has been awful (see attestations). If we go down that route it would probably have to ship with Portable Ruby like zlib already is, though that would still require code to handle system Ruby scenarios.

We could write a pure Ruby library  for Zstd. Though by the time we've done that we might lose most of the benefits of using Zstd.

**Comment by Bo98** [comment]
I guess in theory you only need a decompressor and most magic is done on the compressor side.

File: Library/Homebrew/dev-cmd/bottle.rb [hunk 1]
```diff
@@ -359,18 +378,24 @@ def gnu_tar_formula_ensure_installed_if_needed!
...
         default_tar_args = ["tar", tar_args].freeze
-        return default_tar_args if !args.only_json_tab? || default_tar
+        return default_tar_args unless args.only_json_tab?
 
         # Use gnu-tar as it can be set up for reproducibility better than libarchive
         # and to be consistent between macOS and Linux.
-        gnu_tar_formula = gnu_tar_formula_ensure_installed_if_needed!
-        return default_tar_args if gnu_tar_formula.blank?
+        tar = if system_gnu_tar?
+          "tar"
```
Comment (cho-m) [inline Library/Homebrew/dev-cmd/bottle.rb:390]: Simplest way to update this PR is probably following with additional input parameter for `formula`:
```suggestion
        tar = if formula.name == "gnu-tar" && system_gnu_tar?
          "tar"
```

Though this won't handle case if a user runs `brew bottle gnu-tar` on an old LTS Linux. Seems like a very unlikely situation.

---

A more complete option would be to add OS-specific logic to only check this on Linux-only as macOS doesn't have the same problem in relation to rewritten `ld.so`.

---

Either way, if we do want to consider pure Ruby alternative, then these changes would be a stopgap measure.

Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/bottle.rb:390]: This seems fine to me 👍🏻

**Pull request closed without merging**